src/devices/infactory.c

```diff
@@ -1,84 +1,127 @@
+    int i;
+    char c;
+    char *p = msg;
+    for (i = 0; i < 10; i++) {
+        c = ((b[i >> 1]) >> (4 * (1 - (i & 1)))) & 0x0F;
+        c = (c > 9) ? c + 55 : c + 48;
+        *p++ = c;
+    }
+    *p = 0;
+}
+
+static int infactory_crc_check(uint8_t *b) {
+    int i;
+    uint8_t msg_crc, crc;
+    uint8_t msg[5];
+    memcpy(msg, b, 5);
+    msg_crc = msg[1] >> 4;
+    // for CRC computation, channel bits are at the CRC position(!)
+    msg[1] = (msg[1] & 0x0F) | (msg[4] & 0x0F) << 4;
+    crc = msg[0] >> 4;
+    for (i = 4; i < 36; i++) {
+        crc <<= 1;
+        if (msg[i>>3] & (0x80 >> (i&7))) {
+            crc |= 1; 
+        }
+        if (crc & 0x10) {
+            crc ^= 0x13; // Koopmann 0x9, CCITT-4; FP-4; ITU-T G.704
+        }
+    }
+
+    return (crc == msg_crc);
+}
+
 static int infactory_callback(r_device *decoder, bitbuffer_t *bitbuffer)
 {
     bitrow_t *bb;
     data_t *data;
     uint8_t *b;
-    int id, humidity, temp_raw, channel;
+    int id, battery, temp_raw, humidity, channel;
     float temp_f;
+    char *mic = "ERROR";
+    char raw_msg[11] = ".........."; // 10 hex chars plus 0
 
-    if (bitbuffer->bits_per_row[0] != 4) {
-        return DECODE_ABORT_LENGTH;
-    }
-    if (bitbuffer->bits_per_row[1] != 40) {
+    if (bitbuffer->bits_per_row[0] != 40) {
         return DECODE_ABORT_LENGTH;
     }
     bb = bitbuffer->bb;
 
-    /* Check that 4 bits of preamble is 0 */
+    /* Check that the last 4 bits of message are not 0 (channel number 1 - 3) */
     b = bb[0];
-    if (b[0]&0x0F)
-        return DECODE_ABORT_EARLY;
-
-    /* Check that last 4 bits of message is 0x1 */
-    b = bb[1];
     if (!(b[4]&0x0F))
         return DECODE_ABORT_EARLY;
 
     id       = b[0];
-    humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD
+    battery  = (b[1] >> 2) & 1;
     temp_raw = (b[2] << 4) | (b[3] >> 4);
-    temp_f   = (float)temp_raw * 0.1 - 90;
+    humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD, 'A0'=100%rH
     channel  = b[4] & 0x03;
 
+    temp_f   = (float)temp_raw * 0.1 - 90;
+
+    mic = infactory_crc_check(b) ? "CRC" : "CRC_ERROR"; // we could also return DECODE_FAIL_MIC;
+    infactory_raw_msg(raw_msg, b);                      // DEBUG: print raw message bits
+
     /* clang-format off */
     data = data_make(
             "model",            "",             DATA_STRING, _X("inFactory-TH","inFactory sensor"),
             "id",               "ID",           DATA_INT, id,
             "channel",          "Channel",      DATA_INT, channel,
+            "battery",          "Battery",      DATA_INT, battery,
             "temperature_F",    "Temperature",  DATA_FORMAT, "%.02f °F", DATA_DOUBLE, temp_f,
-            "humidity",         "Humidity",     DATA_FORMAT, "%u %%", DATA_INT, humidity,
+            "humidity",         "Humidity",     DATA_FORMAT, "%u %%rH", DATA_INT, humidity,
+            "mic",              "Integrity",    DATA_STRING, mic,
```

Comment:
mic is always CRC when it is present.

Reply:
Now I understand why it was defined as a string constant in other places.